pytket/phir/phirgen_parallel.py

```diff
@@ -0,0 +1,317 @@
+    )
+    num_scs_per_qubit = {}
+    index = 0
+
+    for qubit in sub_commands:
+        num_scs_per_qubit[qubit] = len(sub_commands[qubit])
+        # set every qubits value to be -4
+        # prevents KeyError in check for group number
+        # will get set to a valid group number the first time the qubit is used
+        qubits2groups[qubit] = -4
+    max_len = max(num_scs_per_qubit.values())
+
+    while index < max_len:
+        for qubit in sub_commands:
+            # check to make sure you are not accessing beyond the end of the list
+            if index < num_scs_per_qubit[qubit]:
+                sc = sub_commands[qubit][index]
+                gate = sc.op.type
+                if gate == tk.OpType.Rz:
+                    group_number = rz_group_number
+                    valid_pll_op = True
+                elif gate == tk.OpType.PhasedX:
+                    group_number = r1xy_group_number
+                    valid_pll_op = True
+                else:
+                    group_number = other_group_number
+                    valid_pll_op = False
+                # does a group exist for that gate type?
+                group_available = group_number in groups
+                # is that group later in execution than the
+                # most recent group for an op on that qubit?
+                order_preserved = group_number > qubits2groups[qubit]
+                # is the group size still under the maximum allowed parallel ops?
+                group_size = len(groups[group_number]) if group_number in groups else 0
+                group_not_too_large = group_size < max_parallel_sq_gates
+                # is the op parallelizable (only RZ or R1XY)?
+                if (
+                    group_available
+                    and order_preserved
+                    and group_not_too_large
+                    and valid_pll_op
+                ):
+                    groups[group_number].append(sc)
+                    qubits2groups[qubit] = group_number
+                else:
+                    # make a new group:
+                    if gate == tk.OpType.Rz:
+                        rz_group_number += 3
+                        group_number = rz_group_number
+                    elif gate == tk.OpType.PhasedX:
+                        r1xy_group_number += 3
+                        group_number = r1xy_group_number
+                    else:
+                        other_group_number += 3
+                        group_number = other_group_number
+                    groups[group_number] = [sc]
+                    qubits2groups[qubit] = group_number
+        index += 1
+
+    return dict(sorted(groups.items()))
+
+
+def groups2qops(groups: dict[int, list[tk.Command]], ops: list[dict[str, Any]]) -> None:  # noqa: PLR0912
+    """Convert the groups of parallel ops to properly formatted PHIR."""
+    for group_number in groups:
+        group = groups[group_number]
```

Comment:
```suggestion
    for group in groups.values():
```
Same as https://github.com/CQCL/pytket-phir/pull/53/files#r1419669378, `process_sub_commands` could just return a list of Shards.

Reply:
process_sub_commands only parallelizes the sub commands within a shard, it does not know anything about the shard structure